tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php

```diff
@@ -0,0 +1,166 @@
+ *
+ * @property BudgetRecommendationQuery|MockObject $budget_recommendation_query
+ * @property ISO3166DataProvider|MockObject       $iso_provider;
+ * @property BudgetRecommendationController       $controller
+ */
+class BudgetRecommendationControllerTest extends RESTControllerUnitTest {
+
+	protected const ROUTE_BUDGET_RECOMMENDATION  = '/wc/gla/ads/campaigns/budget-recommendation';
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->budget_recommendation_query = $this->createMock( BudgetRecommendationQuery::class );
+		$this->budget_recommendation_query->method( 'where' )
+			->willReturn( $this->budget_recommendation_query );
+
+		$this->iso_provider = $this->createMock( ISO3166DataProvider::class );
+		$this->middleware = $this->createMock( Middleware::class );
+
+		$this->controller = new BudgetRecommendationController( $this->server, $this->budget_recommendation_query, $this->middleware );
+		$this->controller->register();
+		$this->controller->set_iso3166_provider( $this->iso_provider );
+	}
+
+	public function test_get_budget_recommendation() {
+		$budget_recommendation_params = [
+			'country_codes' => ['JP', 'TW', 'GB', 'US'],
+		];
+
+		$budget_recommendation_data = [
+			[
+				'country'           => 'US',
+				'daily_budget_low'  => '330',
+				'daily_budget_high' => '930',
+			],
+			[
+				'country'           => 'GB',
+				'daily_budget_low'  => '245',
+				'daily_budget_high' => '625',
+			],
+			[
+				'country'           => 'TW',
+				'daily_budget_low'  => '95',
+				'daily_budget_high' => '255',
+			],
+			[
+				'country'           => 'JP',
+				'daily_budget_low'  => '110',
+				'daily_budget_high' => '320',
+			]
+		];
+
+		$expected_response_data = [
+			'currency'        => 'TWD',
+			'recommendations' => [
+				[
+					'country'           => 'US',
+					'daily_budget_low'  => 330.0,
+					'daily_budget_high' => 930.0,
+				],
+				[
+					'country'           => 'GB',
+					'daily_budget_low'  => 245.0,
+					'daily_budget_high' => 625.0,
+				],
+				[
+					'country'           => 'TW',
+					'daily_budget_low'  => 95.0,
+					'daily_budget_high' => 255.0,
+				],
+				[
+					'country'           => 'JP',
+					'daily_budget_low'  => 110.0,
+					'daily_budget_high' => 320.0,
+				]
+			],
+		];
+
+		$this->middleware->expects( $this->once() )
+			->method( 'get_ads_currency' )
+			->willReturn( 'TWD' );
+
+		$this->budget_recommendation_query->expects( $this->exactly(2) )->method( 'where' );
+
+		$this->budget_recommendation_query->expects( $this->once() )
+			->method( 'get_results' )
+			->willReturn( $budget_recommendation_data );
+
+		$response = $this->do_request( self::ROUTE_BUDGET_RECOMMENDATION, 'GET', $budget_recommendation_params );
+
+		$this->assertSame( $expected_response_data, $response->get_data() );
+		$this->assertEquals( 200, $response->get_status() );
+	}
+
+	public function test_get_budget_recommendation_without_query_parameters() {
```

Comment:
It looks more like we are testing with an empty string instead of without query parameters. If I test the same in postman:

`GET https://domain.test/wp-json/wc/gla/ads/campaigns/budget-recommendation?country_codes=`
I get the error 'Not a valid alpha2 key'

I guess that's because the iso provider is mocked so it doesn't throw the validation error in the test. But it's probably better to focus on the error we expect without sending any parameters (after setting country_codes to required).

If we want to specifically test for the response `Cannot find any budget recommendations`, I think we should create a separate test where we do pass the countries in the request but get an empty response from the query. A scenario like that could happen if the recommendations DB import failed.

Reply:
> It looks more like we are testing with an empty string instead of without query parameters.

Thanks for pointing this out! I updated it in 37e5212, I think this is because I did not set `required => true` in the collection params schema so the built-in validator didn't respond the error 400, also as you said the iso provider was mocked so it didn't respond the correct "invalid alpha2 key" error in the unit test.

> If we want to specifically test for the response Cannot find any budget recommendations, I think we should create a separate test where we do pass the countries in the request but get an empty response from the query. A scenario like that could happen if the recommendations DB import failed.

For this I updated in 57b73b9.